### PR TITLE
Feature/add uniqueness validation to standup meeting group user

### PR DESCRIPTION
**Pull request by wwrk22**
## What's the change?
- Added model-level uniqueness constraint to prevent a user from the joining the same standup meeting group more than once.
- Another PR will implement a database-level constraint by adding a unique index.
- Included a fix for a flash message that wasn’t showing when a user joins a standup meeting group.

## What key workflows are impacted?
N/A

## Highlights / Surprises / Risks / Cleanup
In the future, we should consider the different paths a request can take on the route `/standup_meeting_groups/:standup_meeting_group_id/joins`.
e.g. How do we handle a duplicate POST request to the route caused by the user refreshing the browser during the processing of the original request.

## Demo / Screenshots
Test file `spec/requests/standup_meeting_groups/join_standup_meeting_group_spec.rb` has been written to demonstrate the failure of a request to duplicate a `StandupMeetingGroupUser` record.

## Issue ticket number and link
#251 

## Checklist before requesting a review

Please delete items that are not relevant.

- [x] Did you add appropriate automated tests?
- [x] Did you consider risks around security, performance, etc.?
- [x] Any validations to data needed?
- [x] Related to readability and consistency: Did you add comments and/or documentation? (README, Notion, etc.)
- [x] Is everything elevated to the highest level? (e.g., can logic be moved out of view into controllers, or out of controllers into models?)
- [x] Did you leave helpful inline PR comments for the reviewer(s)?
- [x] Did you include instructions for how to test in the ticket?
- [x] Did you add any relevant tags and decide if this PR is a Draft vs. Ready for Review?
- [x] Do you have a plan for how this change should be rolled back? (e.g., how do you deal with a migration rollback? Is your feature released behind a feature flag?)

### CODE REVIEW

File: app/controllers/standup_meeting_groups/joins_controller.rb [hunk 1]
```diff
@@ -7,12 +8,8 @@ def create
 
     authorize @standup_meeting_group_user, policy_class: StandupMeetingGroup::JoinPolicy
 
-    @standup_meeting_group_user.save!
```
Comment (wwrk22) [inline app/controllers/standup_meeting_groups/joins_controller.rb:10]: According to the [API docs](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-save-21), the only two errors that could be raised by `#save!` are:
- `ActiveRecord::RecordInvalid` due to validation failure
- `ActiveRecord::RecordNotSaved` due to an associated `before_*` callback throwing a `:abort`

If there are existing callbacks that could throw `:abort`, perhaps they could be handled at a higher level, outside the scope of an individual controller since `StandupMeetingGroups::JoinsController` by itself doesn’t seem to be capable of triggering such a scenario.

File: app/controllers/standup_meeting_groups/joins_controller.rb [hunk 2]
```diff
@@ -7,12 +8,8 @@ def create
...
-
-    respond_to do |format|
-      format.turbo_stream do
-        flash.now[:notice] = "You have joined #{@standup_meeting_group.name}!"
-      end
```
Comment (wwrk22) [inline app/controllers/standup_meeting_groups/joins_controller.rb:15]: The `respond_to` block has been removed since
1. A request received at this controller action’s route is automatically accepted as a `turbo_stream` one.
2. We’re not handling any other formats.

File: app/controllers/standup_meeting_groups/joins_controller.rb [hunk 3]
```diff
@@ -7,12 +8,8 @@ def create
...
+    if @standup_meeting_group_user.save
+      flash.now[:notice] = "You have joined #{@standup_meeting_group.name}!"
     end
```
Comment (wwrk22) [inline app/controllers/standup_meeting_groups/joins_controller.rb:13]: Because there is no redirection after the save, it seems to be the case that the user cannot attempt to join the same standup meeting group via the application UI. If there was a redirection, I’d assume that the possibility of the user refreshing the browser during the request needs to be handled. For this reason, the save failure is not explicitly handled for now.

File: app/models/standup_meeting_group_user.rb [hunk 4]
```diff
@@ -18,4 +18,10 @@ class StandupMeetingGroupUser < ApplicationRecord
 
   belongs_to :standup_meeting_group
   belongs_to :user
+
+  validates :user_id,
```
Comment (toyhammered) [inline app/models/standup_meeting_group_user.rb:22]: Can you had a test for this validation in the model?

**Review by toyhammered — CHANGES_REQUESTED** [review]
This looks great! Just one small test I'd like added.

File: db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+class MakeIndexesUniqueForStandupMeetingGroupUsers < ActiveRecord::Migration[7.0]
+  def change
+    # Remove duplicate records.
+    unique_ids = StandupMeetingGroupUser.group(
+      :standup_meeting_group_id, :user_id
+    ).pluck('MIN(id)')
+
+    StandupMeetingGroupUser.where.not(id: unique_ids).delete_all
```
Comment (GALTdea) [inline db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb:8]: @toyhammered Daniel Might want this in a separate task

File: db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class MakeIndexesUniqueForStandupMeetingGroupUsers < ActiveRecord::Migration[7.0]
```
Comment (toyhammered) [inline db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb:1]: All of this needs to be put in a separate PR.
1. We add unique constraint at model level
2. We run rake task to make sure everything is uniq at db level

Then in second PR
1. we add uniq index because it is gaurenteed to work now

**Review by toyhammered — CHANGES_REQUESTED** [review]

File: db/schema.rb [hunk 7]
```diff
@@ -179,8 +179,8 @@
     t.bigint "user_id", null: false
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
-    t.index ["standup_meeting_group_id", "user_id"], name: "index_smg_users_on_smg_id_and_user_id"
-    t.index ["user_id", "standup_meeting_group_id"], name: "index_smg_users_on_user_id_and_smg_id"
+    t.index ["standup_meeting_group_id", "user_id"], name: "index_smg_users_on_smg_id_and_user_id", unique: true
+    t.index ["user_id", "standup_meeting_group_id"], name: "index_smg_users_on_user_id_and_smg_id", unique: true
```
Comment (toyhammered) [inline db/schema.rb:183]: So when you remove a migration the steps need to go:
1. `rake db:rollback` (make sure it is before the migration was run
2. rerun `rake db:migrate`

This will ensure it is removed properly (because you never want to manually touch this file)

**Review by toyhammered — CHANGES_REQUESTED** [review]

File: db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb (hunk 6 above)
Comment (toyhammered) [reply db/migrate/20231114144505_make_indexes_unique_for_standup_meeting_group_users.rb]: sounds good! thank you

File: db/schema.rb (hunk 7 above)
Comment (wwrk22) [reply db/schema.rb]: Ok, I ran the rollback successfully.
Schema is back to the way it was.

**Review by toyhammered — APPROVED** [review]
Once tests pass this looks good!

File: db/schema.rb (hunk 7 above)
Comment (afogel) [reply db/schema.rb]: @wwrk22 I think, as long as you're addressing it in the next PR, it's fine to disable.
To disable, I'd search for the error message first to figure out the specific rubocop to disable, then add a rubocop ignore comment before for that rule and re-enable it after the offending block of code.

Comment (wwrk22) [reply db/schema.rb]: Will do. Thanks for the tip, @afogel !

**Comment by GALTdea** [comment]
test
